core/app/models/spree/price.rb

```diff
@@ -6,13 +6,25 @@ class Price < Spree::Base
 
     belongs_to :variant, -> { with_deleted }, class_name: 'Spree::Variant', touch: true
 
+    after_initialize :valid_from_today!, if: -> { valid_from.blank? }
+
     validate :check_price
     validates :amount, allow_nil: true, numericality: {
       greater_than_or_equal_to: 0,
       less_than_or_equal_to: MAXIMUM_AMOUNT
     }
 
-    after_save :set_default_price
+    scope :valid_before, -> (date) { where("#{self.table_name}.valid_from <= ?", date).order(valid_from: :desc) }
+    scope :currently_valid, -> { valid_before(Time.current) }
+
+    def self.default
+      currently_valid.first
+    end
+
+    # Returns a cache key for all prices in a passed in relation
+    def self.cache_key
+      Digest::MD5.hexdigest(pluck(:id, :updated_at).flatten.join("/"))
```

Comment:
We'll need to order the results from `pluck` somehow to ensure that they're returned in a consistent order.